File: go/cmd/primes.go
package main

import (
	_ "embed"
	"errors"
	"fmt"
	"math/rand"
	"net"
	"net/http"
	"os"
	"os/signal"
	"strconv"
	"syscall"
	"time"
)

const AppName = "Most Inefficient Prime Number Sieve Web Server"

// Single HTML display page that contains client and display code.
//
//go:embed display.html
var page string

// Global HTTP server.
var server http.Server

// Main program sets up web handlers and starts the web server.
func main() {
	fmt.Println("Starting " + AppName)
	defer fmt.Println("Finished " + AppName)

	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		// Main page is in a constant defined below.
		fmt.Println("    PageHandler")
		_, _ = fmt.Fprintf(w, page)
	})
	http.HandleFunc("/sieve/", func(w http.ResponseWriter, r *http.Request) {
		// Sieve runs as server push and uses multiple server-side threads.
		fmt.Println("    SieveHandler")
		// These settings are required for server push.
		w.Header().Set("Content-Type", "text/event-stream")
		w.Header().Set("Cache-Control", "no-cache")

		// Channel for commands to server push handler in browser client.
		c := make(chan string)
		// Thread to pick up commands from channel and send to response writer.
		go commander(w, c)
		// Generate infinite numbers in this handler thread to hold it open.
		generator(c)
	})

	// Getting an unused port is a little complicated.
	// First use net.Listen() to get one.
	var err error
	var listener net.Listener
	if listener, err = net.Listen("tcp", ":0"); err != nil {
		fmt.Printf("!   Unable to create listener port: %s\n", err)
		return
	} else {
		// At this point the port found by the Listener is available.
		fmt.Printf(">   Connect to http://localhost:%d/\n",
			listener.Addr().(*net.TCPAddr).Port)
	}
	// There's no need to configure any fields in the http.Server object.
	server = http.Server{}
	handleInterrupts() // Depends on the existence of the global server object.
	// Use the pre-existing listener to serve web pages.
	if err := server.Serve(listener); err != nil {
		if errors.Is(err, http.ErrServerClosed) {
			fmt.Println("!   Server terminated")
		} else {
			fmt.Printf("*   Error running web server: %s\n", err)
		}
	}
}

// Pull strings from command channel and send to response writer.
// These are server pushed to open connection to client handler.
func commander(w http.ResponseWriter, commands chan string) {
	for {
		cmd := <-commands
		if _, err := w.Write([]byte("data: " + cmd + "\n\n")); err != nil {
			fmt.Println("!   Command channel (browser page) closed")
			_ = server.Close()
			return
		}

		// Flush after every dump.
		if f, ok := w.(http.Flusher); ok {
			f.Flush()
		}
	}
}

// Filter numbers from further up the chain.
// This is run as a separate thread for each prime number 'window'.
// It is called from the number generator and from subsequent filters.
func filter(commands chan string, c chan int, prime int) {
	var o chan int

	// Creating a new filter means a new display box.
	commands <- "make " + strconv.Itoa(prime)

	for {
		// Get the number from the input channel.
		number := <-c
		id := strconv.Itoa(prime)
		// Show the number under generator in the prime box and wait a bit.
		commands <- "generator " + id + " " + strconv.Itoa(number)
		time.Sleep(time.Duration(250+rand.Intn(250)) * time.Millisecond)

		if number%prime == 0 {
			// Number is divisible by this prime, it 'fails' and gets destroyed.
			commands <- "fail " + id
		} else {
			// Number is not divisible, 'success' means going on the chain.
			commands <- "pass " + id
			time.Sleep(time.Duration(250+rand.Intn(250)) * time.Millisecond)
			commands <- "generator " + strconv.Itoa(prime)

			if o == nil {
				// There is no output channel yet, so we have a new prime;
				// make the next filter in the chain.
				o = make(chan int)
				go filter(commands, o, number)
			} else {
				// The next filter is already there, push the number on.
				o <- number
			}
		}
	}
}

// Generate numbers for filtering for prime numbers.
func generator(commands chan string) {
	var c chan int

	for i := 2; ; i++ {
		// Show the new number in the generator box and wait a bit.
		commands <- "gen " + strconv.Itoa(i)
		time.Sleep(time.Duration(250+rand.Intn(250)) * time.Millisecond)

		if c == nil {
			// The first filter must be created here.
			c = make(chan int)
			go filter(commands, c, i)
		} else {
			// The first filter exists, just push the number on the channel.
			c <- i
		}

		commands <- "gen"
	}
}

// Handle user- and system- generated interrupts by closing the HTTP server.
func handleInterrupts() {
	userChannel := make(chan os.Signal)
	signal.Notify(userChannel, os.Interrupt)
	go func() {
		<-userChannel
		// No exlamation point here as program always prints '^C' at the beginning.
		fmt.Println("  User interrupt")
		_ = server.Close()
	}()
	sysChannel := make(chan os.Signal)
	signal.Notify(sysChannel, syscall.SIGTERM)
	go func() {
		<-sysChannel
		fmt.Println("!   System interrupt")
		_ = server.Close()
	}()
}

File: dart/bin/primes.dart
import 'dart:async';
import 'dart:io';
import 'dart:math';

const appName = "Most Inefficient Prime Number Sieve Web Server";

late final HttpServer server;

main() async {
  print("Starting $appName");
  try {
    // No Dart facility for embedding file data into String during build.
    // Load it from a file in a (hopefully) known location.
    List<String> path = Platform.script.toFilePath().split("/");
    path.removeLast();
    String scriptDir = path.join("/");
    String displayPage = await File("$scriptDir/display.html").readAsString();

    // Note: Use io.HttpServer instead of shelf/shelf_io package.
    server = await HttpServer.bind('localhost', 0);
    final commandStream = CommandStream();
    handleInterrupts(
        commandStream); // Depends on the existence of the global server object.
    print("> Connect to http://localhost:${server.port}/");
    await server.forEach((HttpRequest request) async {
      switch (request.uri.path) {
        case "/":
          print("    PageHandler");
          request.response.headers.contentType = ContentType.html;
          request.response.write(displayPage);
          request.response.close();
        case "/sieve/":
          print("    SieveHandler");
          // These settings are required for server push.
          // Note: There is no predefined ContentType for text/event-stream.
          request.response.headers.add("Content-Type", "text/event-stream");
          request.response.headers.add("Cache-Control", "no-cache");
          // Note: Must turn off buffering or the browser doesn't get immediate updates.
          request.response.bufferOutput = false;
          sendCommands(request.response, commandStream.stream);
          generate(commandStream);
        case "/favicon.ico":
          request.response.statusCode = HttpStatus.notFound;
          request.response.close();
        default:
          print("!   Unknown URI path: ${request.uri.path}");
          request.response.statusCode = HttpStatus.badRequest;
          request.response.close();
      }
    });
  } finally {
    print("Finished $appName");
  }
  /* TODO: Despite much experimentation I was unable to get the application to exit.
   * Execution would get to this point and the main routine would exit,
   * but something somewhere (likely an asynchronous wait) was always hanging things up.
   * So finally I decided to use a bigger hammer...
   */
  exit(0);
}

/// Accept commands and add them to the controller stream.
///
/// Does not implement onListen(), onPause(), onResume() or onCancel()
/// as the stream has limited, well-defined usage.
class CommandStream {
  CommandStream() : controller = StreamController();
  final StreamController<String> controller;

  Stream<String> get stream => controller.stream;

  send(String command) {
    controller.add(command);
  }
}

/// Filter numbers from further up the chain.
///
/// A filter is created in the chain for each prime number found.
/// Numbers are passed to the filter for evaluation.
/// Commands are sent to the display page to animate results.
class Filter {
  Filter(this.prime, this.commandStream) {
    // Make a new box in the display for this filter.
    commandStream.send("make $prime");
  }
  final int prime;
  final CommandStream commandStream;
  Filter? next;

  static const int visibilityWait = 250; // milliseconds
  static final random = Random();

  /// Wait a bit to allow browser animation to be visible to the user.
  Future<void> wait() {
    return Future.delayed(Duration(
        milliseconds: visibilityWait + random.nextInt(visibilityWait)));
  }

  evaluate(int number) async {
    // Show the number in the filter box on the browser.
    commandStream.send("eval $prime $number");
    await wait();

    // Evaluate the number to see if the filter's prime divides evenly.
    if (number % prime == 0) {
      commandStream.send("fail $prime");
    } else {
      commandStream.send("pass $prime");
      await wait();
      commandStream.send("clear $prime");

      if (next != null) {
        next!.evaluate(number);
      } else {
        next = Filter(number, commandStream);
      }
    }
  }
}

void generate(CommandStream commandStream) async {
  int i = 2;
  Filter? filter;
  while (true) {
    commandStream.send("gen $i");
    await Future.delayed(Duration(seconds: 1));

    filter ??= Filter(i, commandStream);
    filter.evaluate(i++);

    commandStream.send("gen");
  }
}

/// Capture interrupts and stop service.
void handleInterrupts(CommandStream cmdStream) {
  ProcessSignal.sigint.watch().listen((signal) {
    print("  User interrupt");
    server.close(force: true);
  });
  ProcessSignal.sigterm.watch().listen((signal) {
    print("!   System interrupt");
    server.close(force: true);
  });
  /* Note: Android Studio uses SIGKILL to shut down Dart applications.
   * It is not possible to call watch() on ProcessSignal.sigkill.
   */
}

/// Pull command strings from the specified stream and write them to the open response writer.
void sendCommands(HttpResponse response, Stream<String> commands) async {
  await for (final command in commands) {
    response.write("data: $command\n\n");
    var info = response.connectionInfo;
    await response.flush();
  }
}
